ra/ra.go

```diff
@@ -1036,22 +1072,40 @@ func (ra *RegistrationAuthorityImpl) issueCertificateInner(
 		return emptyCert, err
 	}
 
+	var authzs map[string]*core.Authorization
 	// If the orderID is 0 then this is a classic issuance and we need to check
 	// that the account is authorized for the names in the CSR.
 	if oID == 0 {
-		err = ra.checkAuthorizations(ctx, names, account.ID)
+		authzs, err = ra.checkAuthorizations(ctx, names, account.ID)
 	} else {
 		// Otherwise, if the orderID is not 0 we need to follow the order based
 		// issuance process and check that this specific order is fully authorized
 		// and associated with the expected account ID
-		err = ra.checkOrderAuthorizations(ctx, names, acctID, oID)
+		authzs, err = ra.checkOrderAuthorizations(ctx, names, acctID, oID)
 	}
 	if err != nil {
 		// Pass through the error without wrapping it because the called functions
 		// return BoulderError and we don't want to lose the type.
 		return emptyCert, err
 	}
 
+	// Collect up a certificateRequestAuthz that stores the ID and challenge type
+	// of each of the valid authorizations we used for this issuance.
+	logEventAuthzs := make(map[string]certificateRequestAuthz, len(names))
+	for name, authz := range authzs {
+		var solvedByChallengeType string
+		// If the authz has no solved by challenge type there has been an internal
+		// consistency violation worth raising an internal server error about.
+		if solvedByChallengeType = authz.SolvedBy(); solvedByChallengeType == "" {
+			return emptyCert, berrors.InternalServerError("Authz %q has status %q but empty SolvedBy()", authz.ID, authz.Status)
```

Comment:
This will happen sometimes when we hit the magic validation race, which we currently hide with some WFE code, which will trigger a 500 now. Are we okay with that?

Reply:
@jsha What do you think? We could set the `solvedByChallengeType` to `""` and log a warning instead.

Reply:
I think we should avoid returning a 500 in this case, but we should log it.

Reply:
:+1: addressed in 2390ccd